dep0/lib/03_typecheck/src/type_assign.cpp

```diff
@@ -0,0 +1,332 @@
+    };
+    auto func = type_assign(ctx, app.func.get());
+    if (not func)
+        return std::move(func.error());
+    auto func_type = [&] () -> expected<expr_t::pi_t>
+    {
+        if (auto* const type = std::get_if<expr_t>(&func->properties.sort.get()))
+        {
+            ast::beta_delta_normalize(ctx.delta_reduction_context(), *type);
+            if (auto const pi = std::get_if<expr_t::pi_t>(&type->value))
+                return *pi;
+        }
+        std::ostringstream err;
+        pretty_print(err << "cannot invoke expression of type `", func->properties.sort.get()) << '`';
+        return error(err.str());
+    }();
+    if (not func_type)
+        return std::move(func_type.error());
+    if (func_type->args.size() != app.args.size())
+    {
+        std::ostringstream err;
+        err << "passed " << app.args.size() << " arguments but was expecting " << func_type->args.size();
+        return error(err.str());
+    }
+    std::vector<expr_t> args;
+    for (auto const i: std::views::iota(0ul, func_type->args.size()))
+    {
+        auto arg = check_expr(ctx, app.args[i], func_type->args[i].type);
+        if (not arg)
+            return std::move(arg.error());
+        if (func_type->args[i].var)
+            substitute<properties_t>(
+                *func_type->args[i].var,
+                *arg,
+                func_type->args.begin() + i + 1,
+                func_type->args.end(),
+                func_type->ret_type.get(),
+                nullptr);
+        args.push_back(std::move(*arg));
+    }
+    return make_legal_expr(std::move(func_type->ret_type.get()), expr_t::app_t{std::move(*func), std::move(args)});
+}
+
+expected<expr_t> type_assign_abs(
+    context_t const& ctx,
+    parser::expr_t::abs_t const& f,
+    source_loc_t const& location,
+    std::optional<source_text> const& name)
+{
+    auto f_ctx = ctx.extend();
+    auto func_type = check_pi_type(f_ctx, f.args, f.ret_type.get());
+    if (not func_type)
+        return std::move(func_type.error());
+    // if a function has a name it can call itself recursively;
+    // to typecheck recursive functions we need to add them to the current function context before checking the body
+    if (name)
+    {
+        auto const func_name = expr_t::var_t{*name};
+        if (auto ok = f_ctx.try_emplace(func_name, location, make_legal_expr(*func_type, func_name)); not ok)
+            return error_t::from_error(std::move(ok.error()));
+    }
+    auto [arg_types, ret_type] = std::get<expr_t::pi_t>(func_type->value);
+    auto state = proof_state_t(f_ctx, ret_type.get());
+    auto body = check_body(state, f.body);
+    if (not body)
+        return std::move(body.error());
+    // so far so good, but we now need to make sure that all branches contain a return statement,
+    // with the only exception of functions returning `unit_t` because the return statement is optional;
+    if (not returns_from_all_branches(*body))
```

Comment:
Now that we have initializer-lists, we could make `{}` type-check against `unit_t` and make `return` statements always mandatory